ResearchKit/ActiveTasks/ORKToneAudiometryStepViewController.m

```diff
@@ -0,0 +1,210 @@
+
+    [self.toneAudiometryContentView.tapButton addTarget:self action:@selector(buttonPressed:forEvent:) forControlEvents:UIControlEventTouchDown];
+
+    self.currentTestIndex = 0;
+    self.testingFrequencies = @[@125, @250, @500, @1000, @2000, @4000, @8000];
+    self.audioGenerator = [ORKAudioGenerator new];
+}
+
+- (void)viewDidAppear:(BOOL)animated {
+    [super viewDidAppear:animated];
+
+    [self start];
+}
+
+- (ORKStepResult *)result {
+    ORKStepResult *sResult = [super result];
+
+    // "Now" is the end time of the result, which is either actually now,
+    // or the last time we were in the responder chain.
+    NSDate *now = sResult.endDate;
+
+    NSMutableArray *results = [NSMutableArray arrayWithArray:sResult.results];
+
+    ORKToneAudiometryResult *toneResult = [[ORKToneAudiometryResult alloc] initWithIdentifier:(NSString *__nonnull)self.step.identifier];
+    toneResult.startDate = sResult.startDate;
+    toneResult.endDate = now;
+    toneResult.samples = [self.samples copy];
+
+    [results addObject:toneResult];
+    sResult.results = [results copy];
+
+    return sResult;
+}
+
+- (void)stepDidFinish {
+    [super stepDidFinish];
+
+    self.expired = YES;
+    [self.toneAudiometryContentView finishStep:self];
+    [self goForward];
+}
+
+- (void)start {
+    [super start];
+
+    [self startCurrentTest];
+}
+
+- (void)suspend {
+    [super suspend];
+    // ...
+}
+
+- (void)resume {
+    [super resume];
+    // ...
+}
+
+- (void)finish {
+    [super finish];
+    // ...
+}
+
+- (IBAction)buttonPressed:(id)button forEvent:(UIEvent *)event {
+    if (self.samples == nil) {
+        _samples = [NSMutableArray array];
+    }
+
+    ORKToneAudiometrySample *sample = [ORKToneAudiometrySample new];
+    NSUInteger frequencyIndex = self.currentTestIndex/2;
+    NSNumber *frequency = self.testingFrequencies[frequencyIndex];
+    sample.frequency = frequency;
+    sample.channel = (self.currentTestIndex%2 == 0) ? ORKAudioChannelLeft : ORKAudioChannelRight;
+    sample.amplitude = @(self.audioGenerator.volumeAmplitude);
+
+    [self.samples addObject:sample];
+
+    [NSRunLoop cancelPreviousPerformRequestsWithTarget:self selector:@selector(testExpired) object:nil];
+    [self.audioGenerator stop];
+
+    [self startNextTestOrFinish];
+}
+
+- (void)testExpired {
+    [self.audioGenerator stop];
+
+    [self startNextTestOrFinish];
+}
+
+- (void)startNextTestOrFinish {
+    self.currentTestIndex ++;
+    if (self.currentTestIndex == (self.testingFrequencies.count * 2)) {
+        [self finish];
+    }
+    else {
+        [self startCurrentTest];
+    }
+}
+
+- (ORKToneAudiometryStep *)toneAudiometryStep {
+    return (ORKToneAudiometryStep *)self.step;
+}
+
+- (void)startCurrentTest {
+    const NSTimeInterval SoundDuration = self.toneAudiometryStep.toneDuration;
+
+    NSUInteger frequencyIndex = self.currentTestIndex/2;
+    if (frequencyIndex >= self.testingFrequencies.count) {
+        return;
```

Comment:
If you actually hit this case in testing, it would be indicative of a bug, correct? Should this be an assert?